port/linux/omrvmem.c

```diff
@@ -1076,61 +1091,90 @@ get_hugepages_info(struct OMRPortLibrary *portLibrary, vmem_hugepage_info_t *pag
 	return 1;
 }
 
+static BOOLEAN
+set_flags_for_mmap(int *flags)
+{
+	BOOLEAN useBackingFile = FALSE;
+#if defined(MAP_ANONYMOUS)
+	*flags |= MAP_ANONYMOUS | MAP_PRIVATE;
+#elif defined(MAP_ANON)
+	*flags |= MAP_ANON | MAP_PRIVATE;
+#else
+	useBackingFile = TRUE;
+	*flags |= MAP_PRIVATE;
+#endif
+	return useBackingFile;
+}
+
 static void *
-default_pageSize_reserve_memory(struct OMRPortLibrary *portLibrary, void *address, uintptr_t byteAmount, struct J9PortVmemIdentifier *identifier, uintptr_t mode, uintptr_t pageSize, OMRMemCategory *category)
+reserve_memory_with_mmap(struct OMRPortLibrary *portLibrary, void *address, uintptr_t byteAmount, struct J9PortVmemIdentifier *identifier, uintptr_t mode, uintptr_t pageSize, OMRMemCategory *category)
 {
 	/* This function is cloned in J9SourceUnixJ9VMem (omrvmem_reserve_memory).
 	 * Any changes made here may need to be reflected in that version.
 	 */
 	int fd = -1;
-	int ft = -1;
 	int flags = 0;
 	void *result = NULL;
 	int protectionFlags = PROT_NONE;
-	BOOLEAN useBackingSharedFile = FALSE;
 	BOOLEAN useBackingFile = FALSE;
+	BOOLEAN useBackingSharedFile = FALSE;
+	BOOLEAN areHugePagesEnabled = FALSE;
 
-	Trc_PRT_vmem_default_reserve_entry(address, byteAmount);
+	Trc_PRT_vmem_reserve_entry(address, byteAmount);
 
-	if(mode & OMRPORT_VMEM_MEMORY_MODE_SHARE_FILE_OPEN) {
+	if (0 != (mode & OMRPORT_VMEM_MEMORY_MODE_SHARE_FILE_OPEN)) {
 		flags |= MAP_SHARED;
 		useBackingSharedFile = TRUE;
 	} else {
-#if defined(MAP_ANONYMOUS)
-		flags |= MAP_ANONYMOUS | MAP_PRIVATE;
-		useBackingSharedFile = FALSE;
-#elif defined(MAP_ANON)
-		flags |= MAP_ANON | MAP_PRIVATE;
-		useBackingSharedFile = FALSE;
-#else
-		useBackingFile = TRUE;
-		flags |= MAP_PRIVATE;
-#endif
+		useBackingFile = set_flags_for_mmap(&flags);
+	}
+
+#if defined(MAP_HUGETLB)
+	if (0 != (mode & OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES)) {
+		flags |= MAP_HUGETLB;
+		areHugePagesEnabled = TRUE;
 	}
+#endif /* MAP_HUGETLB */
 
 	if (useBackingSharedFile) {
-		char filename[FILE_NAME_SIZE + 1];
-		sprintf(filename, "omrvmem_temp_%zu_%09d_%zu",  (size_t)omrtime_current_time_millis(portLibrary),
+		memfd_call memfd_createDL = (memfd_call)PPG_memfd_createDL;
+		if (NULL != memfd_createDL) {
+			Trc_PRT_vmem_reserve_doubleMapAPI_available(address, byteAmount);
+			mode |= OMRPORT_VMEM_MEMORY_MODE_DOUBLE_MAP_AVAILABLE;
+			int ft = -1;
+			uintptr_t fdFlags = 0;
+			char filename[FILE_NAME_SIZE + 1];
+			sprintf(filename, "omrvmem_temp_%zu_%09d_%zu",  (size_t)omrtime_current_time_millis(portLibrary),
 								getpid(),
 								(size_t)pthread_self());
-		fd = shm_open(filename, O_RDWR | O_CREAT, 0600);
-		shm_unlink(filename);
-		ft = ftruncate(fd, byteAmount);
-
-		if (fd == -1 || ft == -1) {
-			Trc_PRT_vmem_default_reserve_failed(address, byteAmount);
-			update_vmemIdentifier(identifier, NULL, NULL, 0, 0, 0, 0, 0, NULL, -1);
-			Trc_PRT_vmem_default_reserve_exit(result, address, byteAmount);
-			return result;
+#if defined(MFD_HUGETLB)
```

Comment:
Is this defined on the OpenJ9 compile platforms?

Reply:
`MFD_HUGETLB` is a `memfd_create` flag and exists since glibc 2.27. Not sure if that answers the question either. But this is a fairly new syscall, and most, if not all, of our platforms doesn't have that flag.

Reply:
```
#if defined(LINUX)
uintptr_t mfd_hugeTBL = 0;
#if defined(MFD_HUGETLB)
    mfd_hugeTBL = MFD_HUGETLB;
#else
    /* hard coded value of MFD_HUGETLB */
    mfd_hugeTBL = 4U;
#endif
#endif
```

Reply:
Oh I see makes sense. Should I still go ahead and do the same for `MAP_HUGETLB`, or just for `MFD_HUGETLB`?

Reply:
And should we do the way you proposed or something like:
```
#if !defined(MFD_HUGETLB)
#define MFD_HUGETLB 0x4
#endif
```

Even though `MFD_HUGETLB` is only used in one place

Reply:
Sure, your proposal is fine with me (although I'm not an OMR committer). 

> Should I still go ahead and do the same for MAP_HUGETLB, or just for MFD_HUGETLB?

I don't think it's necessary, all the minimum platforms should have it. I'd use `#if defined(LINUX) and assume it's available, not `#if defined(MAP_HUGETLB)`. I think the LINUX check is important, in particular if defining MFD_HUGETLB, to avoid the constant being used on other platform in the future (the value may be different).

Reply:
Cool, the `#if defined(LINUX)` might be redundant in this case since this file is only compiled in Linux platform, right? If that's the case then I'll go ahead and remove all `#if defined(MAP_HUGETLB)` and update `MFD_HUGETLB`

Reply:
Sounds good.